app/code/community/Bolt/Boltpay/Helper/Api.php

```diff
@@ -1182,4 +1193,83 @@ public function getMagentoUrl($route = '', $params = array()){
         }
         return Mage::getUrl($route, $params);
     }
+
+
+    /**
+     *  Updates the shipping address data and, if necessary, the billing address data to the Magento
+     *  quote
+     *
+     * @param Mage_Sales_Model_Quote    $quote             The quote to which the address will be applied
+     * @param array                     $shippingAddress   The Bolt formatted address data
+     *
+     * @return  array   The shipping address applied in Magento compatible format
+     */
+    public function applyShippingAddressToQuote( $quote, $shippingAddress ) {
+
+        $directory = Mage::getModel('directory/region')->loadByName($shippingAddress->region, $shippingAddress->country_code);
+        $region = $directory->getName(); // For region field should be the name not a code.
+        $regionId = $directory->getRegionId(); // This is require field for calculation: shipping, shopping price rules and etc.
+
+        $addressData = array(
+            'email' => $shippingAddress->email ?: $shippingAddress->email_address,
+            'firstname' => $shippingAddress->first_name,
+            'lastname' => $shippingAddress->last_name,
+            'street' => $shippingAddress->street_address1 . ($shippingAddress->street_address2 ? "\n" . $shippingAddress->street_address2 : ''),
+            'company' => $shippingAddress->company,
+            'city' => $shippingAddress->locality,
+            'region' => $region,
+            'region_id' => $regionId,
+            'postcode' => $shippingAddress->postal_code,
+            'country_id' => $shippingAddress->country_code,
+            'telephone' => $shippingAddress->phone ?: $shippingAddress->phone_number
+        );
+
+        if ($quote->getCustomerId()) {
+            $customerSession = Mage::getSingleton('customer/session');
+            $customerSession->setCustomerGroupId($quote->getCustomerGroupId());
+            $customer = Mage::getModel("customer/customer")->load($quote->getCustomerId());
+            $address = $customer->getPrimaryShippingAddress();
+
+            if (!$address) {
+                $address = Mage::getModel('customer/address');
+
+                $address->setCustomerId($customer->getId())
+                    ->setCustomer($customer)
+                    ->setIsDefaultShipping('1')
+                    ->setSaveInAddressBook('1')
+                    ->save();
+
+
+                $address->addData($addressData);
+                $address->save();
+
+                $customer->addAddress($address)
+                    ->setDefaultShippingg($address->getId())
+                    ->save();
+            }
+        }
+        $quote->removeAllAddresses();
```

Comment:
do you remember why we remove address here? if so can you add comment explaining why?

Reply:
@daisy1754 One reason is because we need this address to be clean and reflect only what the customer enters from the BoltCheckout modal.  Some address values can been set by other plugins and typically by the standard Magento shopping cart `Estimate Shipping and Tax` module.